ddtrace/settings/_inferred_base_service.py

```diff
@@ -145,39 +156,51 @@ def detect_service(args: List[str]) -> Optional[str]:
-            detector_name = detector_instance.name
-
-            if detector_instance.matches(command):
-                detectors.update({detector_name: detector_instance})
-                # append to a list of arg indexes to ignore since they are executables
-                executable_args.add(i)
-                continue
-            elif _is_executable(command):
-                # append to a list of arg indexes to ignore since they are executables
-                executable_args.add(i)
-
-    args_to_search = []
-    for i, arg in enumerate(args):
-        # skip any executable args
-        if i not in executable_args:
-            args_to_search.append(arg)
-
-    # Iterate through the matched detectors
-    for detector in detectors.values():
-        metadata = detector.detect(args_to_search)
-        if metadata and metadata.name:
-            CACHE[cache_key] = metadata.name
-            return metadata.name
+    try:
+        # Check both the included command args as well as the executable being run
+        possible_commands = [*args, sys.executable]
+        executable_args = set()
+
+        # List of detectors to try in order
+        detectors = {}
+        for detector_class in detector_classes:
+            detector_instance = detector_class(dict(os.environ))
+
+            for i, command in enumerate(possible_commands):
+                detector_name = detector_instance.name
+
+                if detector_instance.matches(command):
+                    detectors.update({detector_name: detector_instance})
+                    # append to a list of arg indexes to ignore since they are executables
+                    executable_args.add(i)
+                    continue
+                elif _is_executable(command):
+                    # append to a list of arg indexes to ignore since they are executables
+                    executable_args.add(i)
+
+        args_to_search = []
+        for i, arg in enumerate(args):
+            # skip any executable args
+            if i not in executable_args:
+                args_to_search.append(arg)
+
+        # Iterate through the matched detectors
+        for detector in detectors.values():
+            metadata = detector.detect(args_to_search)
+            if metadata and metadata.name:
+                CACHE[cache_key] = metadata.name
+                return metadata.name
+
+        # Iterate through the matched detectors again, this time not skipping args preceded by flag args
+        for detector in detectors.values():
+            metadata = detector.detect(args_to_search, skip_args_preceded_by_flags=False)
+            if metadata and metadata.name:
+                CACHE[cache_key] = metadata.name
+                return metadata.name
+    except Exception as ex:
+        # Catch any unexpected errors to be extra safe
+        log.debug("Unexpected error during inferred base service detection: ", ex)
+
     CACHE[cache_key] = None
     return None
```

Comment:
Do we want None or default-python-service or something else?

Reply:
we fallback to `default-python-service` for V1, None for V0, that logic is within the service naming schema files.